<?xml version="1.0"?>
<launch>
    <arg name="vocabularty_path" default="$(find orb_slam2_cuda)/Vocabulary/ORBvoc.txt" />
    <arg name="camera_setting_path" default="$(find my_pipeline)/calibration/TUM1_rect.yaml" />
    <arg name="bUseViewer" default="true" />
    <arg name="bEnablePublishROSTopic" default="true" />


    <arg name="cam_name" default="openni_camera" />

    <node pkg="rosbag" type="play" name="player" output="screen" args="--clock -r 0.1 /root/TUM_office_recording.bag " launch-prefix="bash -c 'sleep 7; $0 $@' "/>


   <!-- <node name="orb_slam2_cuda" pkg="orb_slam2_cuda" type="Mono" output="screen"
    	args="$(arg vocabularty_path) $(arg camera_setting_path) $(arg bUseViewer) $(arg bEnablePublishROSTopic)">
    	<remap from="/camera/image_raw" to="/camera/rgb/image_rect"/>
    	<remap from="/transform" to="/transform_rgb_frame"/>
  	</node> -->

    <!-- from flame nodelet launch -->
    <!--
    <arg name="keyframes" default="keyframes" />

    <arg name="namespace" default="flame" />
    <arg name="manager" default="manager" />
    <arg name="standalone" default="true" />
    <arg name="output_dest" default="screen" />

    <arg name="load_params" default="true" />
    <arg name="param_file" default="flame_nodelet.yaml" />

    <arg name="bag_data" default="false" />

     <!- - <node pkg="tf2_ros" type="static_transform_publisher" name="body_to_camera"
            args="0.0 0.0 0.0 0 0 0 0 body $(arg cam_name)" >
      </node> - ->


    <group ns="$(arg namespace)" >


      <arg name="world_frame_id" default="world" />
      <arg name="camera_world_frame_id" default="camera_world" />
          <node pkg="tf2_ros" type="static_transform_publisher" name="camera_world_to_world"
               args="0.0 0.0 0.0 0.5 -0.5 0.5 -0.5 $(arg world_frame_id) $(arg camera_world_frame_id)" >
          </node>


        <!- - Launch the nodelet manager. - ->
        <node if="$(arg standalone)"
              pkg="nodelet" type="nodelet" name="$(arg manager)"
              args="manager" output="$(arg output_dest)" >
        </node>

        <!- - Launch the nodelet.. - ->
        <node pkg="nodelet" type="nodelet" name="flame_nodelet"
              args="load flame_ros/FlameNodelet $(arg manager)"
              output="$(arg output_dest)" launch-prefix="bash -c 'sleep 7; $0 $@' ">
          <rosparam command="load" if="$(arg load_params)" file="$(find my_pipeline)/cfg/$(arg param_file)"/>
          <remap from="/image" to="/camera/rgb/image_rect" />
         <remap from="/info" to="/camera/rgb/camera_info" />
          <remap from="/transform" to="/orb_slam2/transform"/>
          <remap from="keyframes" to="$(arg keyframes)" />
        </node>

    </group>

    <node name="rvizfiestavisualisation" pkg="rviz" type="rviz" output="log" args="-d $(find my_pipeline)/cfg/flame.rviz" />
-->

  <!-- size of map, change the size in x, y, z according to your application -->
  <arg name="map_size_x" value="50.0"/>
  <arg name="map_size_y" value="50.0"/>
  <arg name="map_size_z" value=" 10.0"/>

  <!-- topic of your odometry such as VIO or LIO -->

    <node pkg="tf" type="static_transform_publisher" name="tf_53" args="0 0 0 0 0 0 world navigation 100" />

  <!-- main algorithm params -->
  <include file="$(find my_pipeline)/launch/algorithm_esdf.xml">

    <arg name="map_size_x_" value="$(arg map_size_x)"/>
    <arg name="map_size_y_" value="$(arg map_size_y)"/>
    <arg name="map_size_z_" value="$(arg map_size_z)"/>

    <!-- sensor pose: transform of camera frame in the world frame -->
    <arg name="sensor_pose_topic" value="/transform_rgb_frame"/>
    <!-- depth topic: depth image, 640x480 by default -->
    <!-- cloud topic: point cloud measurement  -->
    <!-- subscribe ONLY TO ONE of the two topics -->
    <arg name="depth_topic" value="/camera/depth_registered/image_rect"/>
    <arg name="semantic_topic" value="/semantic"/>
    <!--<arg name="cloud_topic" value="/pcl_render_node/cloud"/> -->

    <!-- intrinsic params of the depth camera -->
    <arg name="cx" value="318.643040"/>
    <arg name="cy" value="255.313989"/>
    <arg name="fx" value="517.306408"/>
    <arg name="fy" value="516.469215"/>
    <!-- maximum velocity and acceleration the drone will reach -->
    <arg name="max_vel" value="2.0" />
    <arg name="max_acc" value="2.0" />
  </include>

    <node name="rvizesdfvisualisation" pkg="rviz" type="rviz" output="log" args="-d $(find my_pipeline)/cfg/semantic_esdf.rviz" />
    <node name="segmentator" pkg="tdnet_nyud" type="segmentation.py" output="screen">
       <remap from="image_raw" to="/camera/rgb/image_rect_color" />
    </node>

   <!-- <node pkg="fiesta" type="test_fiesta" name="fiesta" output="screen"
          required="true">

        <param name="resolution" value="0.1"/>
        <param name="update_esdf_every_n_sec" value="0.1"/>

        <!- -Hash Table only- ->
        <param name="reserved_size" value="1000000"/>

        <!- -Array only- ->
        <param name="lx" value="-10.0"/>
        <param name="ly" value="-10.0"/>
        <param name="lz" value="-1"/>
        <param name="rx" value="+10.0"/>
        <param name="ry" value="+10.0"/>
        <param name="rz" value="+3"/>

        <!- - raycasting parameters - ->
        <param name="min_ray_length" value="0.5"/>
        <param name="max_ray_length" value="5.0"/>
        <!- -If hash table is used, make sure the value is 0. In array implementation, 0 means no other thread.- ->
        <param name="ray_cast_num_thread" value="0"/>



        <!- - needed when using depth image- ->
        <!- -        <param name="center_x" value="323.120483"/>- ->
        <!- -        <param name="center_y" value="236.743210"/>- ->
        <!- -        <param name="focal_x" value="385.754486"/>- ->
        <!- -        <param name="focal_y" value="385.754486"/>- ->
            <param name="center_x" value="318.643040"/>
            <param name="center_y" value="255.313989"/>
            <param name="focal_x" value="517.306408"/>
            <param name="focal_y" value="516.469215"/>

        <!- - probabilistic grid map - ->
        <param name="p_hit" value="0.70"/>
        <param name="p_miss" value="0.35"/>
        <param name="p_min" value="0.12"/>
        <param name="p_max" value="0.97"/>
        <param name="p_occ" value="0.80"/>

        <!- - global / local - ->
        <param name="global_map" value="true"/>
        <param name="global_update" value="true"/>
        <param name="global_vis" value="true"/>
        <param name="radius_x" value="3.0"/>
        <param name="radius_y" value="3.0"/>
        <param name="radius_z" value="1.5"/>

        <!- -depth_filter - ->
        <param name="use_depth_filter" value="true"/>
        <param name="depth_filter_tolerance" value="0.1"/>
        <param name="depth_filter_max_dist" value="10.0"/>
        <param name="depth_filter_min_dist" value="0.1"/>
        <!- -unit: pixel- ->
        <param name="depth_filter_margin" value="0"/>

        <!- - visulization - ->
        <!- -0 for no visulize- ->
        <param name="visualize_every_n_updates" value="1"/>
        <param name="slice_vis_max_dist" value="2.0"/>
        <!- - relative to the lz if array is used, relative to the origin if hash table is used - ->
        <param name="slice_vis_level" value="1.6"/>
        <param name="vis_lower_bound" value="0"/>
        <param name="vis_upper_bound" value="+10"/>

        <!- - subsribe source - ->
        <remap from="~depth" to="/camera/depth_registered/image_rect"/>
        <remap from="~transform" to="/transform_rgb_frame"/>
    </node>

    <node name="rvizvisualisation" pkg="rviz" type="rviz" output="log" args="-d $(find fiesta)/demo.rviz" />
-->

</launch>
